packages/opentelemetry-plugin-user-interaction/src/userInteraction.ts

```diff
@@ -0,0 +1,452 @@
+    this._unpatchHistoryApi();
+
+    shimmer.wrap(history, 'replaceState', this._patchHistoryMethod());
+    shimmer.wrap(history, 'pushState', this._patchHistoryMethod());
+    shimmer.wrap(history, 'back', this._patchHistoryMethod());
+    shimmer.wrap(history, 'forward', this._patchHistoryMethod());
+    shimmer.wrap(history, 'go', this._patchHistoryMethod());
+  }
+
+  /**
+   * Patches the certain history api method
+   */
+  _patchHistoryMethod() {
+    const plugin = this;
+    return (original: any) => {
+      return function patchHistoryMethod(this: History, ...args: unknown[]) {
+        const url = `${location.pathname}${location.hash}${location.search}`;
+        const result = original.apply(this, args);
+        const urlAfter = `${location.pathname}${location.hash}${location.search}`;
+        if (url !== urlAfter) {
+          plugin._updateInteractionName(urlAfter);
+        }
+        return result;
+      };
+    };
+  }
+
+  /**
+   * unpatch the history api methods
+   */
+  _unpatchHistoryApi() {
+    if (isWrapped(history.replaceState))
+      shimmer.unwrap(history, 'replaceState');
+    if (isWrapped(history.pushState)) shimmer.unwrap(history, 'pushState');
+    if (isWrapped(history.back)) shimmer.unwrap(history, 'back');
+    if (isWrapped(history.forward)) shimmer.unwrap(history, 'forward');
+    if (isWrapped(history.go)) shimmer.unwrap(history, 'go');
+  }
+
+  /**
+   * Updates interaction span name
+   * @param url
+   */
+  _updateInteractionName(url: string) {
+    const span: types.Span | undefined = this._tracer.getCurrentSpan();
+    if (span && typeof span.updateName === 'function') {
+      span.updateName(`${EVENT_NAVIGATION_NAME} ${url}`);
+    }
+  }
+
+  /**
+   * Patches zone cancel task - this is done to be able to correctly
+   * decrement the number of remaining tasks
+   */
+  private _patchZoneCancelTask() {
+    const plugin = this;
+    return (original: any) => {
+      return function patchCancelTask<T extends Task>(
+        this: Zone,
+        task: AsyncTask
+      ) {
+        const currentZone = Zone.current;
+        const currentSpan = currentZone.get(ZONE_SCOPE_KEY);
+        if (currentSpan && plugin._shouldCountTask(task, currentZone)) {
+          plugin._decrementTask(currentSpan);
+        }
+        return original.call(this, task) as T;
+      };
+    };
+  }
+
+  /**
+   * Patches zone schedule task - this is done to be able to correctly
+   * increment the number of tasks running within current zone but also to
+   * save time in case of timeout running from xhr plugin when waiting for
+   * main request from PerformanceResourceTiming
+   */
+  private _patchZoneScheduleTask() {
+    const plugin = this;
+    return (original: any) => {
+      return function patchScheduleTask<T extends Task>(
+        this: Zone,
+        task: AsyncTask
+      ) {
+        const currentZone = Zone.current;
+        const currentSpan: types.Span = currentZone.get(ZONE_SCOPE_KEY);
```

Comment:
`types.Span` is optional, right?

Reply:
what you mean ?

Reply:
I think he is asking if `currentZone.get` always returns a Span or can it also return null or undefined?